### PR TITLE
Refactor `httpx/_utils.py`

### PR DESCRIPTION
**Pull request by yezz123**
Based on what @tomchristie recommended to make https://github.com/encode/httpx/pull/1909 this PR into a shorter one to validate multiple things and let it more clear and ready for review.

### CODE REVIEW

File: httpx/_utils.py [hunk 1]
```diff
@@ -457,19 +457,18 @@ def __init__(self, pattern: str) -> None:
...
-                # *example.com should match "www.example.com" and "example.com"
-                domain = re.escape(url.host[1:])
-                self.host_regex = re.compile(f"^(.+\\.)?{domain}$")
-            else:
-                # example.com should match "example.com" but not "www.example.com"
-                domain = re.escape(url.host)
-                self.host_regex = re.compile(f"^{domain}$")
+            # example.com should match "example.com" but not "www.example.com"
+            domain = re.escape(url.host)
+            self.host_regex = re.compile(f"^{domain}$")
```
Comment (tomchristie) [inline httpx/_utils.py:471]: I like the refactoring you've done on this block, yup. I think it makes enough of a readability difference that it's worth having in.

File: httpx/_utils.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ def guess_json_utf(data: bytes) -> typing.Optional[str]:
         if sample[1::2] == _null2:  # 2nd and 4th are null
             return "utf-16-le"
         # Did not detect 2 valid UTF-16 ascii-range characters
-    if nullcount == 3:
+    elif nullcount == 3:
```
Comment (tomchristie) [inline httpx/_utils.py:123]: Either we should update all the cases here that could be `elif`, or none of them.
We're not really getting any value from the change, so I'd suggest we just leave it.

```suggestion
    if nullcount == 3:
```

File: httpx/_utils.py [hunk 3]
```diff
@@ -23,7 +23,7 @@
     {chr(c): "%{:02X}".format(c) for c in range(0x1F + 1) if c != 0x1B}
 )
 _HTML5_FORM_ENCODING_RE = re.compile(
-    r"|".join([re.escape(c) for c in _HTML5_FORM_ENCODING_REPLACEMENTS.keys()])
+    r"|".join([re.escape(c) for c in _HTML5_FORM_ENCODING_REPLACEMENTS])
```
Comment (tomchristie) [inline httpx/_utils.py:26]: I could go either way on this one. Possibly it's more obvious with `.keys()` what the effect is. Possibly not. But, okay, sure.

**Comment by tomchristie** [comment]
Okay, I've reverted one `elif` change that was just dangling on its own.
The other two changes make sense.

**Comment by yezz123** [comment]
> Okay, I've reverted one `elif` change that was just dangling on its own. The other two changes make sense.

@tomchristie that's great, if there is any other things we can discuss about it and fix it 🚀

File: httpx/_utils.py (hunk 3 above)
Comment (tomchristie) [inline httpx/_utils.py:26]:
```suggestion
    r"|".join([re.escape(c) for c in _HTML5_FORM_ENCODING_REPLACEMENTS.keys()])
```

**Comment by tomchristie** [comment]
I've reverted the minor change to `_HTML5_FORM_ENCODING_RE` - it's unrelated to the `elif` block changes.

Given that I think *those* are a clear win, let's *just pull those in* as a single nicely scoped pull request.

**Review by tomchristie — APPROVED** [review]
Nice little improvement to the flow there, thank you!

**Pull request merged**